hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java

```diff
@@ -908,6 +909,7 @@ public HRegion(final HRegionFileSystem fs, final WAL wal, final Configuration co
           conf.getBoolean(HConstants.ENABLE_CLIENT_BACKPRESSURE,
               HConstants.DEFAULT_ENABLE_CLIENT_BACKPRESSURE);
 
+    this.hasPrefix = conf.get(KeyPrefixRegionSplitRestriction.PREFIX_LENGTH_KEY) != null;
```

Comment:
This assumes we'll have to re-create the HRegion anytime we modify the prefix. Given that modifications to the prefix require a major compaction to take effect, I thought this would be okay.

Reply:
This is the only assumption that I'm fuzzy on. I'm not sure whether major compactions would trigger reconstruction of HRegion objects comprehensively

This leads to two questions in my mind:
* could we just fetch the variable live on each call to `isValidSplit`
* do we definitely need this qualification at all? On one hand it's nice to keep changes as targeted as possible, but I also wonder whether these 0->0 splits could just never be valid

Reply:
> could we just fetch the variable live on each call to isValidSplit

I had a similar thought. The primary reason for caching this var is to avoid the unnecessary call to `Configuration#get` if possible.

I did a bit more digging and realized that the underlying `KeyPrefixRegionSplitRestriction` also stores the prefix length, and is cached by the [HRegion](https://github.com/HubSpot/hbase/blob/23239e7eefbacd0c05d0f4a2b3f11e8e52862de3/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java#L1068). It might actually make sense to update `hasPrefix` anytime `splitRestriction` is updated. 

Another way to do this check is to infer whether there's a prefix based on the result of [splitRestriction.getRestrictedSplitPoint](https://github.com/HubSpot/hbase/blob/23239e7eefbacd0c05d0f4a2b3f11e8e52862de3/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java#L8877). 

If the ret value is shorter than the initial ret value, we know there's a prefix

```
int preSplitLength = ret.length;
ret = splitRestriction.getRestrictedSplitPoint(ret);

if (ret.length != preSplitLength) {
  // Prefix exists
} else {
  // No prefix
}
```

Which might be a cleaner way of keeping "hasPrefix" up to date without explicitly querying the configuration.

> do we definitely need this qualification at all? On one hand it's nice to keep changes as targeted as possible, but I also wonder whether these 0->0 splits could just never be valid

This is a fair point, I think the code path would work regardless, I was mostly attempting to avoid doing any necessary work if possible, but it's possible the optimization isn't necessary, especially if it means we have to manage some more state.

Reply:
Hmm, you've actually changed my mind. If we aren't explicitly checking for a prefix length, I don't think we should care about a prefix at all. A split that will generate an empty region should never be allowed to surface, making those changes.